src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt

```diff
@@ -129,6 +132,20 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
     }
   }
 
+  override fun loadNewChatId(callback: () -> Unit) {
+    id = null
+    promptPanel.textArea.isEnabled = false
+    promptPanel.textArea.emptyText.text = "Connecting to agent..."
+    ApplicationManager.getApplication().executeOnPooledThread {
+      getInitializedServer(project).thenAccept { server ->
+        id = server.chatNew().get()
+        promptPanel.textArea.isEnabled = true
+        promptPanel.textArea.emptyText.text = "Ask a question about this code..."
+        callback.invoke()
```

Comment:
Also, it makes assumption on which thread callback should be called.
Since you cannot control there what will be in the callback code (e.g. long operation vs UI update) running it in proper theread should be responsibility of the callback code, not the caller.

Reply:
The callback is to be called having the new `id` assigned. We need to do things in the following order:
1.  Reset `id` (set `id` to `null`)
2. Update UI with the connecting state (`text = "Connecting to agent..."`)
3. With an initialized server send a request for the new chat (that cannot be called on UI thread) 
4. Having the new `id`, update UI with the ready state (`emptyText.text = "Ask a question about this code..."`)
5. Run callback

Reply:
I reviewed the threads run here. The UI actions are called from AWT. The other actions are on pooled threads.